File: src/forager/core/game.py
from __future__ import annotations
from dataclasses import dataclass
from enum import Enum, auto
from pathlib import Path

from forager.core.paths import games_dir

GENERIC_CONTAINERS = {
    "standalone", "series", "minecraft", "steam", "rpgmaker", "rpg",
    "games", "instances", "launcher", "single", "flash", "drm-free",
}

ENGINE_NAMES = {"other", "rpgmaker", "unity", "unreal"}


class Source(Enum):
    STEAM = auto()
    MINECRAFT = auto()
    STANDALONE = auto()


@dataclass
class Game:
    name: str
    source: Source
    path: Path | None = None
    app_id: str | None = None
    launch_cmd: list[str] | None = None
    sort_key: str | None = None
    search_names: list[str] | None = None
    installed: bool = True

    def __hash__(self):
        return hash((self.source, self.app_id or str(self.path)))

    def __eq__(self, other):
        if not isinstance(other, Game):
            return NotImplemented
        return (self.source, self.app_id or str(self.path)) == (
            other.source,
            other.app_id or str(other.path),
        )

    @property
    def source_name(self) -> str:
        return {
            Source.STEAM: "Steam",
            Source.MINECRAFT: "Minecraft",
            Source.STANDALONE: "Standalone",
        }[self.source]

    @property
    def display_path(self) -> str:
        """Path shown in the UI: relative to the games directory starting at
        the holder folder (e.g. ``drm-free/series/…``), else the absolute
        path.  Owned-but-uninstalled games have no path and show a dash."""
        if self.path is None:
            return "—"
        try:
            rel = self.path.resolve().relative_to(games_dir())
            return "/".join(rel.parts)
        except ValueError:
            return str(self.path)

    @property
    def sgdb_search(self) -> tuple[list[str], str] | None:
        """SGDB search plan: (queries, match_term), or None to skip search.

        Searches the holding (series) folder rather than the leaf folder name,
        so e.g. ``series/sequel/asylum`` searches ``sequel``. Returns None for
        generic container folders (minecraft, standalone, ...) to avoid wrong
        matches. Single games under an engine folder (``other``, ``unity``, …)
        fall back to searching by their own name. ``search_names`` always wins
        when set.
        """
        if self.search_names:
            return (list(self.search_names), "")
        if self.path is None:
            return None
        if self.source == Source.STEAM:
            return None
        try:
            parts = list(self.path.resolve().relative_to(games_dir()).parts)
        except ValueError:
            return None
        if len(parts) >= 2 and parts[-2].lower() in ENGINE_NAMES:
            return ([parts[-1]], "")
        while len(parts) >= 2 and parts[-2].lower() in GENERIC_CONTAINERS:
            parts.pop()
        if len(parts) >= 2:
            return ([parts[-2]], parts[-1])
        return None

File: src/core/Game.cpp
#include "core/Game.h"
#include "core/Paths.h"

#include <QSet>
#include <QDir>

static const QSet<QString> GENERIC_CONTAINERS = {
    "standalone", "series", "minecraft", "steam", "rpgmaker", "rpg",
    "games", "instances", "launcher", "single", "flash", "drm-free",
};

static const QSet<QString> ENGINE_NAMES = {"other", "rpgmaker", "unity", "unreal"};

Game::Game(QString name, Source source)
    : m_name(std::move(name))
    , m_source(source)
{
}

QString Game::identifier() const
{
    if (!m_app_id.isEmpty()) return m_app_id;
    if (!m_path.isEmpty()) return m_path;
    return m_name;
}

std::optional<std::pair<QStringList, QString>> Game::sgdbSearch() const
{
    if (!m_search_names.isEmpty())
        return std::make_pair(m_search_names, QString());
    if (m_path.isEmpty()) return std::nullopt;
    if (m_source == Source::Steam) return std::nullopt;

    QString gamesRoot = paths::gamesDir();
    if (gamesRoot.isEmpty()) return std::nullopt;

    QString canonical = QFileInfo(m_path).canonicalFilePath();
    if (canonical.isEmpty()) canonical = m_path;
    if (!canonical.startsWith(gamesRoot)) return std::nullopt;

    QString rel = canonical.mid(gamesRoot.length());
    if (rel.startsWith('/')) rel.remove(0, 1);
    if (rel.isEmpty()) return std::nullopt;

    QStringList parts = rel.split('/', Qt::SkipEmptyParts);
    if (parts.size() >= 2 && ENGINE_NAMES.contains(parts[parts.size() - 2].toLower()))
        return std::make_pair(QStringList{parts.last()}, QString());

    while (parts.size() >= 2 && GENERIC_CONTAINERS.contains(parts[parts.size() - 2].toLower()))
        parts.removeAt(parts.size() - 2);

    if (parts.size() >= 2)
        return std::make_pair(QStringList{parts[parts.size() - 2]}, parts.last());
    return std::nullopt;
}

QString Game::displayPath() const
{
    if (m_path.isEmpty()) return QStringLiteral("\u2014");

    QFileInfo info(m_path);
    QString canonical = info.canonicalFilePath();
    if (canonical.isEmpty()) return m_path;

    QString gamesRoot = paths::gamesDir();
    if (canonical.startsWith(gamesRoot)) {
        QString rel = canonical.mid(gamesRoot.length());
        if (rel.startsWith('/')) rel.remove(0, 1);
        return rel;
    }
    return m_path;
}

bool Game::operator==(const Game& other) const
{
    return m_source == other.m_source && identifier() == other.identifier();
}
